File: dsa03_Sorting/mergeSort.go
package main

import (
	"fmt"
)

func main() {
	fmt.Println("Merge Sort")
	arr := []int{3, 5, 8, 9, 6, 2, 1, -1}
	fmt.Println(arr)
	mergeSort(arr, 0, len(arr)-1)
	fmt.Println(arr)
}

func mergeSort(arr []int, left, right int) {
	if left < right {
		mid := int((left + right) / 2)
		mergeSort(arr, left, mid)
		mergeSort(arr, mid+1, right)
		merge(arr, left, mid, right)
	}
}

func merge(arr []int, left, mid, right int) {
	i, j := left, mid+1

	temp := []int{}

	for i <= mid && j <= right {
		if arr[i] <= arr[j] {
			temp = append(temp, arr[i])
			i++
		} else {
			temp = append(temp, arr[j])
			j++
		}
	}

	if i <= mid {
		temp = append(temp, arr[i:mid+1]...)
	}

	if j <= right {
		temp = append(temp, arr[j:right+1]...)
	}

	for i, v := range temp {
		arr[i+left] = v
	}
}

File: dsa03_Sorting/mergesort.py
def mergesort(A, left, right):
    if left < right:
        mid = (left + right) // 2

        mergesort(A, left, mid)
        mergesort(A, mid + 1, right)

        merge(A, left, mid, right)


def merge(A, left, mid, right):
    i, j = left, mid + 1
    B = []

    while i <= mid and j <= right:
        if A[i] < A[j]:
            B.append(A[i])
            i += 1
        else:
            B.append(A[j])
            j += 1

    if i <= mid:
        B.extend(A[i:mid + 1])

    if j <= right:
        B.extend(A[j:right + 1])

    for x, v in enumerate(B):
        A[left + x] = v


if __name__ == '__main__':
    A = [3, 5, 8, 9, 6, 2]
    print(A)
    mergesort(A, 0, len(A) - 1)
    print(A)
